File: parsers/thingiverse-parser.py
import requests
import json
from datetime import datetime
import os
import random
import sys
import time

# --- Credentials ---
APP_TOKEN = os.environ.get("THINGIVERSE_TOKEN", "53dba3cff3fbbf0506e34d7fa855f40e")
THINGIVERSE_API = 'https://api.thingiverse.com'

JOKE_KEYWORDS = [
    "joke", "meme", "funny", "gag", "prank", "fake", "parody", "satire", "ironic", "not real", "fake part",
    "keychain", "logo", "decoration", "figurine", "statue", "ornament", "fan art", "toy", "miniature", "sign",
    "display", "stand", "desktop", "accessory", "charm", "pendant", "wall art", "poster", "non-functional"
]

FUNCTIONAL_KEYWORDS = [
    "gear", "knob", "handle", "bracket", "clip", "button", "lever", "mount", "adapter", "joint", "wheel",
    "shaft", "seal", "gasket", "spring", "latch", "hinge", "cap", "plug", "cover", "base", "housing", "shell",
    "replacement", "repair", "fix", "part", "impeller", "pulley", "bushing", "nozzle"
]

SPARE_PARTS_QUERIES = {
    'Bosch': ['Bosch+spare+part', 'Bosch+repair'],
    'Dyson': ['Dyson+spare+part', 'Dyson+repair'],
    'Ikea': ['Ikea+spare+part', 'Ikea+repair'],
    'Samsung': ['Samsung+spare+part', 'Samsung+repair'],
    'LG': ['LG+spare+part', 'LG+repair'],
    'Whirlpool': ['Whirlpool+spare+part', 'Whirlpool+repair'],
    'Philips': ['Philips+spare+part', 'Philips+repair'],
    'Braun': ['Braun+spare+part', 'Braun+repair'],
    'Miele': ['Miele+spare+part', 'Miele+repair'],
    'Xiaomi': ['Xiaomi+spare+part', 'Xiaomi+repair'],
    'Electrolux': ['Electrolux+spare+part', 'Electrolux+repair'],
    'Kenwood': ['Kenwood+spare+part', 'Kenwood+repair'],
    'KitchenAid': ['KitchenAid+spare+part', 'KitchenAid+repair'],
    'DeLonghi': ['DeLonghi+spare+part', 'DeLonghi+repair'],
    'Tefal': ['Tefal+spare+part', 'Tefal+repair'],
    'Beko': ['Beko+spare+part', 'Beko+repair'],
    'General Parts': ['spare+part', 'repair', 'replacement+part', 'fix']
}

HOBBY_QUERIES = {
    'Tabletop': ['dnd', 'warhammer', 'miniature', 'terrain'],
    'Games': ['minecraft', 'pokemon', 'zelda', 'star+wars'],
    'Toys': ['toy', 'puzzle', 'lego'],
    'Decor': ['decoration', 'vase', 'art']
}

AUTO_QUERIES = {
    'Toyota': ['Toyota+part', 'Toyota+accessory'],
    'BMW': ['BMW+part', 'BMW+accessory'],
    'Mercedes': ['Mercedes+part', 'Mercedes+repair'],
    'Audi': ['Audi+part', 'Audi+repair'],
    'Volkswagen': ['VW+part', 'VW+repair'],
    'Tesla': ['Tesla+part', 'Tesla+accessory'],
    'Accessories': ['car+holder', 'cupholder', 'key+fob']
}

HOME_QUERIES = {
    'Makita': ['Makita+spare+part', 'Makita+repair'],
    'Karcher': ['Karcher+spare+part', 'Karcher+repair'],
    'DeWalt': ['DeWalt+part', 'DeWalt+adapter'],
    'Garden': ['garden+tool', 'hose+connector'],
    'Kitchen': ['organizer', 'hook', 'shelf']
}

def is_joke(name, description, mode):
    text = (name + " " + (description or "")).lower()
    for word in JOKE_KEYWORDS:
        if word in text:
            return True
    if mode == 'spare-parts':
        brands = ['bosch', 'dyson', 'samsung', 'lg', 'whirlpool', 'miele', 'ikea', 'kitchenaid']
        if any(b in text for b in brands):
            if not any(fk in text for fk in FUNCTIONAL_KEYWORDS):
                return True
    return False

def estimate_volume(name):
    name_l = name.lower()
    if any(x in name_l for x in ['small', 'tiny', 'knob', 'button', 'clip', 'gear']):
        return round(random.uniform(2, 10), 2)
    if any(x in name_l for x in ['large', 'big', 'housing', 'case', 'box', 'mount']):
        return round(random.uniform(80, 250), 2)
    return round(random.uniform(15, 60), 2)

def fetch_results(queries, mode, headers, existing_ids, max_pages=1, sort='relevant'):
    models = {}
    for category, terms in queries.items():
        for term in terms:
            print(f"[{mode}] Searching '{category}' for: '{term}' (sort={sort})...")
            for page in range(1, max_pages + 1):
                url = f'{THINGIVERSE_API}/search/{term}?sort={sort}&per_page=40&page={page}'
                try:
                    response = requests.get(url, headers=headers, timeout=10)
                    response.raise_for_status()
                    data = response.json()
                except Exception as e:
                    print(f"  - Error: {e}")
                    break

                hits = data.get('hits', [])
                if not hits: break

                found_new_on_page = False
                hit_existing = False
                for item in hits:
                    model_id = f"thingiverse_{item['id']}"
                    
                    if model_id in existing_ids:
                        if sort == 'newest':
                            hit_existing = True
                            # In newest sort, once we hit an existing ID, we can stop this term entirely
                        continue

                    if not item.get('preview_image'): continue
                    if is_joke(item['name'], item.get('description'), mode): continue

                    brand = None
                    if mode == 'spare-parts' and category != 'General Parts':
                        for b in SPARE_PARTS_QUERIES.keys():
                            if b != 'General Parts' and b.lower() in item['name'].lower():
                                brand = b
                                break
                        if not brand: brand = category

                    found_new_on_page = True
                    models[model_id] = {
                        'objectID': model_id,
                        'name': item['name'],
                        'description': item.get('description', ''),
                        'brand': brand,
                        'category': category,
                        'mode': mode,
                        'source': 'thingiverse',
                        'source_url': item['public_url'],
                        'license': item.get('license'),
                        'author': item['creator']['name'] if item.get('creator') else 'Unknown',
                        'image': item['preview_image'],
                        'popularity': item.get('likes_count', 0),
                        'downloads': item.get('download_count', 0),
                        'stl_url': None, # To be fetched lazily
                        'volume_cm3': estimate_volume(item['name']),
                        'indexed_at': datetime.now().isoformat()
                    }

                if hit_existing and sort == 'newest':
                    print(f"  - Hit existing model, stopping search for '{term}'")
                    break

                if not found_new_on_page and max_pages == 1:
                    break

            # Rate limiting friendly delay
            time.sleep(0.5)

    return models

def parse_thingiverse():
    is_full = '--full' in sys.argv
    is_initial = '--initial' in sys.argv

    # Logic:
    # - Initial: Relevant sort, multiple pages
    # - Update: Newest sort, stop when existing hit
    sort = 'relevant' if (is_initial or is_full) else 'newest'
    max_pages = 5 if (is_initial or is_full) else 1

    headers = {'Authorization': f'Bearer {APP_TOKEN}'}
    output_path = 'data/models-index.json'
    os.makedirs(os.path.dirname(output_path), exist_ok=True)

    existing_models = {}
    if os.path.exists(output_path):
        try:
            with open(output_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                for m in data:
                    existing_models[m['objectID']] = m
            print(f"Loaded {len(existing_models)} existing models.")
        except Exception as e:
            print(f"Error loading existing models: {e}")

    existing_ids = set(existing_models.keys())

    all_new_models = {}
    for queries, mode in [
        (SPARE_PARTS_QUERIES, 'spare-parts'),
        (HOBBY_QUERIES, 'hobby'),
        (AUTO_QUERIES, 'auto'),
        (HOME_QUERIES, 'home')
    ]:
        new_items = fetch_results(queries, mode, headers, existing_ids, max_pages=max_pages, sort=sort)
        all_new_models.update(new_items)

    new_count = 0
    for model_id, model_data in all_new_models.items():
        if model_id not in existing_models:
            existing_models[model_id] = model_data
            new_count += 1

    models_list = list(existing_models.values())
    # Always keep it sorted by popularity for the main view
    models_list.sort(key=lambda x: x.get('popularity', 0), reverse=True)

    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(models_list, f, ensure_ascii=False, indent=2)

    print(f"Update complete. Added {new_count} new models. Total: {len(models_list)}")

if __name__ == '__main__':
    parse_thingiverse()

File: parsers/thingiverse-parser.js
import fetch from 'node-fetch';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const APP_TOKEN = process.env.THINGIVERSE_TOKEN || "53dba3cff3fbbf0506e34d7fa855f40e";
const THINGIVERSE_API = 'https://api.thingiverse.com';

const JOKE_KEYWORDS = [
    "joke", "meme", "funny", "gag", "prank", "fake", "parody", "satire", "ironic", "not real", "fake part",
    "keychain", "logo", "decoration", "figurine", "statue", "ornament", "fan art", "toy", "miniature", "sign",
    "display", "stand", "desktop", "accessory", "charm", "pendant", "wall art", "poster", "non-functional"
];

const FUNCTIONAL_KEYWORDS = [
    "gear", "knob", "handle", "bracket", "clip", "button", "lever", "mount", "adapter", "joint", "wheel",
    "shaft", "seal", "gasket", "spring", "latch", "hinge", "cap", "plug", "cover", "base", "housing", "shell",
    "replacement", "repair", "fix", "part", "impeller", "pulley", "bushing", "nozzle"
];

const SPARE_PARTS_QUERIES = {
    'Bosch': ['Bosch+spare+part', 'Bosch+repair'],
    'Dyson': ['Dyson+spare+part', 'Dyson+repair'],
    'Ikea': ['Ikea+spare+part', 'Ikea+repair'],
    'Samsung': ['Samsung+spare+part', 'Samsung+repair'],
    'LG': ['LG+spare+part', 'LG+repair'],
    'Whirlpool': ['Whirlpool+spare+part', 'Whirlpool+repair'],
    'Philips': ['Philips+spare+part', 'Philips+repair'],
    'Braun': ['Braun+spare+part', 'Braun+repair'],
    'Miele': ['Miele+spare+part', 'Miele+repair'],
    'Xiaomi': ['Xiaomi+spare+part', 'Xiaomi+repair'],
    'Electrolux': ['Electrolux+spare+part', 'Electrolux+repair'],
    'Kenwood': ['Kenwood+spare+part', 'Kenwood+repair'],
    'KitchenAid': ['KitchenAid+spare+part', 'KitchenAid+repair'],
    'DeLonghi': ['DeLonghi+spare+part', 'DeLonghi+repair'],
    'Tefal': ['Tefal+spare+part', 'Tefal+repair'],
    'Beko': ['Beko+spare+part', 'Beko+repair'],
    'General Parts': ['spare+part', 'repair', 'replacement+part', 'fix']
};

const HOBBY_QUERIES = {
    'Tabletop': ['dnd', 'warhammer', 'miniature', 'terrain'],
    'Games': ['minecraft', 'pokemon', 'zelda', 'star+wars'],
    'Toys': ['toy', 'puzzle', 'lego'],
    'Decor': ['decoration', 'vase', 'art']
};

const AUTO_QUERIES = {
    'Toyota': ['Toyota+part', 'Toyota+accessory'],
    'BMW': ['BMW+part', 'BMW+accessory'],
    'Mercedes': ['Mercedes+part', 'Mercedes+repair'],
    'Audi': ['Audi+part', 'Audi+repair'],
    'Volkswagen': ['VW+part', 'VW+repair'],
    'Tesla': ['Tesla+part', 'Tesla+accessory'],
    'Accessories': ['car+holder', 'cupholder', 'key+fob']
};

const HOME_QUERIES = {
    'Makita': ['Makita+spare+part', 'Makita+repair'],
    'Karcher': ['Karcher+spare+part', 'Karcher+repair'],
    'DeWalt': ['DeWalt+part', 'DeWalt+adapter'],
    'Garden': ['garden+tool', 'hose+connector'],
    'Kitchen': ['organizer', 'hook', 'shelf']
};

function isJoke(name, description, mode) {
    const text = (name + " " + (description || "")).toLowerCase();
    if (JOKE_KEYWORDS.some(word => text.includes(word))) return true;

    if (mode === 'spare-parts') {
        const brands = ['bosch', 'dyson', 'samsung', 'lg', 'whirlpool', 'miele', 'ikea', 'kitchenaid'];
        if (brands.some(b => text.includes(b.toLowerCase()))) {
            if (!FUNCTIONAL_KEYWORDS.some(fk => text.includes(fk.toLowerCase()))) {
                return true;
            }
        }
    }
    return false;
}

function estimateVolume(name) {
    const nameL = name.toLowerCase();
    if (['small', 'tiny', 'knob', 'button', 'clip', 'gear'].some(x => nameL.includes(x))) {
        return Math.round((Math.random() * 8 + 2) * 100) / 100;
    }
    if (['large', 'big', 'housing', 'case', 'box', 'mount'].some(x => nameL.includes(x))) {
        return Math.round((Math.random() * 170 + 80) * 100) / 100;
    }
    return Math.round((Math.random() * 45 + 15) * 100) / 100;
}

async function fetchResults(queries, mode, headers, existingIds, maxPages = 1, sort = 'relevant') {
    const models = {};
    for (const [category, terms] of Object.entries(queries)) {
        for (const term of terms) {
            console.log(`[${mode}] Searching '${category}' for: '${term}' (sort=${sort})...`);
            for (let page = 1; page <= maxPages; page++) {
                const url = `${THINGIVERSE_API}/search/${term}?sort=${sort}&per_page=40&page=${page}`;
                try {
                    const response = await fetch(url, { headers, timeout: 10000 });
                    if (!response.ok) break;
                    const data = await response.json();

                    const hits = data.hits || [];
                    if (hits.length === 0) break;

                    let foundNewOnPage = false;
                    let hitExisting = false;

                    for (const item of hits) {
                        const modelId = `thingiverse_${item.id}`;

                        if (existingIds.has(modelId)) {
                            if (sort === 'newest') hitExisting = true;
                            continue;
                        }

                        if (!item.preview_image) continue;
                        if (isJoke(item.name, item.description, mode)) continue;

                        let brand = null;
                        if (mode === 'spare-parts' && category !== 'General Parts') {
                            for (const b of Object.keys(SPARE_PARTS_QUERIES)) {
                                if (b !== 'General Parts' && item.name.toLowerCase().includes(b.toLowerCase())) {
                                    brand = b;
                                    break;
                                }
                            }
                            if (!brand) brand = category;
                        }

                        foundNewOnPage = true;
                        models[modelId] = {
                            objectID: modelId,
                            name: item.name,
                            description: item.description || '',
                            brand: brand,
                            category: category,
                            mode: mode,
                            source: 'thingiverse',
                            source_url: item.public_url,
                            license: item.license,
                            author: item.creator ? item.creator.name : 'Unknown',
                            image: item.preview_image,
                            popularity: item.likes_count || 0,
                            downloads: item.download_count || 0,
                            stl_url: null,
                            volume_cm3: estimateVolume(item.name),
                            indexed_at: new Date().toISOString()
                        };
                    }

                    if (hitExisting && sort === 'newest') {
                        console.log(`  - Hit existing model, stopping search for '${term}'`);
                        break;
                    }

                    if (!foundNewOnPage && maxPages === 1) break;
                } catch (e) {
                    console.error(`  - Error: ${e.message}`);
                    break;
                }
                // Small delay to be nice to API
                await new Promise(r => setTimeout(r, 500));
            }
        }
    }
    return models;
}

async function parseThingiverse() {
    const isFull = process.argv.includes('--full');
    const isInitial = process.argv.includes('--initial');

    const sort = (isInitial || isFull) ? 'relevant' : 'newest';
    const maxPages = (isInitial || isFull) ? 5 : 1;

    const headers = { 'Authorization': `Bearer ${APP_TOKEN}` };
    const outputPath = path.resolve(__dirname, '../data/models-index.json');

    let existingModels = {};
    if (fs.existsSync(outputPath)) {
        try {
            const data = JSON.parse(fs.readFileSync(outputPath, 'utf-8'));
            data.forEach(m => existingModels[m.objectID] = m);
            console.log(`Loaded ${Object.keys(existingModels).length} existing models.`);
        } catch (e) {
            console.error(`Error loading existing models: ${e.message}`);
        }
    }

    const existingIds = new Set(Object.keys(existingModels));
    const queryGroups = [
        [SPARE_PARTS_QUERIES, 'spare-parts'],
        [HOBBY_QUERIES, 'hobby'],
        [AUTO_QUERIES, 'auto'],
        [HOME_QUERIES, 'home']
    ];

    let newCount = 0;
    for (const [queries, mode] of queryGroups) {
        const newItems = await fetchResults(queries, mode, headers, existingIds, maxPages, sort);
        for (const [id, data] of Object.entries(newItems)) {
            if (!existingModels[id]) {
                existingModels[id] = data;
                newCount++;
            }
        }
    }

    const modelsList = Object.values(existingModels);
    modelsList.sort((a, b) => (b.popularity || 0) - (a.popularity || 0));

    fs.writeFileSync(outputPath, JSON.stringify(modelsList, null, 2));
    console.log(`Update complete. Added ${newCount} new models. Total: ${modelsList.length}`);
}

parseThingiverse();
